examples/firewall/firewall.c

```diff
@@ -0,0 +1,384 @@
+                return 0;
+        }
+
+        switch (rule) {
+                case 0:
+                        meta->action = ONVM_NF_ACTION_TONF;
+                        meta->destination = destination;
+                        stats.pkt_accept++;
+                        stats.pkt_total++;
+                        if (debug) RTE_LOG(INFO, APP, "Packet from source IP %u has been accepted.\n", ipv4_hdr->src_addr);
+                        break;
+                default:
+                        meta->action = ONVM_NF_ACTION_DROP;
+                        stats.pkt_drop++;
+                        stats.pkt_total++;
+                        if (debug) RTE_LOG(INFO, APP, "Packet from source IP %u has been dropped.\n",ipv4_hdr->src_addr);
+                        break;
+        }
+
+        return 0;
+}
+
+static int lpm_setup(struct onvm_fw_rule **rules, int num_rules) {
+        int i, status, ret;
+        char name[64];
+
+        firewall_req = (struct lpm_request *) rte_malloc(NULL, sizeof(struct lpm_request), 0);
+
+        if (!firewall_req) return 0;
+
+        snprintf(name, sizeof(name), "fw%d-%"PRIu64, rte_lcore_id(), rte_get_tsc_cycles());
+        firewall_req->max_num_rules = 1024;
+        firewall_req->num_tbl8s = 24;
+        firewall_req->socket_id = rte_socket_id();
+        strcpy(firewall_req->name, name);
+
+        status = onvm_nflib_request_lpm(firewall_req);
+
+        if (status < 0) {
+                rte_exit(EXIT_FAILURE, "Cannot get lpm region for firewall\n");
+        }
+
+        lpm_tbl = rte_lpm_find_existing(name);
+
+        if (lpm_tbl == NULL) {
+                printf("No existing LPM_TBL\n");
+        }
+
+        for (i = 0; i < num_rules; ++i) {
+                printf("RULE %d: { ip: %d, depth: %d, action: %d }\n", i, rules[i]->src_ip, rules[i]->depth,
+                       rules[i]->action);
+                ret = rte_lpm_add(lpm_tbl, rules[i]->src_ip, rules[i]->depth, rules[i]->action);
+                if (ret < 0) {
+                        printf("ERROR ADDING RULE %d\n", ret);
+                        return 1;
+                }
+        }
+
+        return 0;
+}
+
+static void lpm_teardown(struct onvm_fw_rule **rules, int num_rules) {
+        int i;
+
+        if (rules) {
+                for (i = 0; i < num_rules; ++i) {
+                        if (rules[i]) free(rules[i]);
+                }
+                free(rules);
+        }
+
+        if (firewall_req) {
+                free(firewall_req);
+        }
+
+        if (lpm_tbl) {
+                rte_lpm_free(lpm_tbl);
+        }
+}
+
+struct onvm_fw_rule **setup_rules(int *total_rules, char *rules_file) {
+        int ip, num_rules;
+        int i = 0;
+        struct onvm_fw_rule **rules;
+
+        cJSON *rules_json = onvm_config_parse_file(rules_file);
+        cJSON *rules_ip = NULL;
+        cJSON *depth = NULL;
+        cJSON *action = NULL;
+
+        if (rules_json == NULL) {
```

Comment:
Okay so what are we trying to resolve here again? I remember we had a brief discussion about this.
The execution of the NF should be along the lines of this:
```sh
cd firewall
./go.sh 1 -d 2 -f rules.json
```
or 
```sh
cd ~/openNetVM/examples
./start_nf.sh firewall 1 -d 2 -f firewall/rules.json
```
or
```sh
cd ~/
 ~/openNetVM-dev/examples/firewall/start_nf.sh firewall  1 -d 2 -f ~/openNetVM-dev/examples/firewall/rules.json
```

Reply:
Running from examples makes the function construct the examples/firewall string and look for the rules there. We could add into the readme that you should try those different paths. For now it does that extra step of checking if we’re running from the examples directory. Do you think it’d be cleaner to assume the rules will always be in the firewall directory (keeping the code as it is) or should we remove that extra directory construction part and have the readme show 3 different options on how to run it?

Reply:
When you run the NF it's assumed that you are giving a path to the rules.json file from the directory you run it from. If it works with the examples I wrote in the comment abovr, then remove the extra step. Looking for the rules file in 2 different places is error prone.

Reply:
that works